File: contests/lc/1161.MaximumLevelSumofaBinaryTree/solution.py
from collections import deque
from typing import Optional


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val: int = 0, left: "Optional[TreeNode]" = None, right: "Optional[TreeNode]" = None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    def maxLevelSum(self, root: Optional[TreeNode]) -> int:
        if root is None:
            return 0

        queue = deque([root])
        level = 0
        best_level = 1
        best_sum = root.val

        while queue:
            level += 1
            level_sum = 0
            for _ in range(len(queue)):
                node = queue.popleft()
                level_sum += node.val
                if node.left is not None:
                    queue.append(node.left)
                if node.right is not None:
                    queue.append(node.right)

            if level_sum > best_sum or level == 1:
                best_sum = level_sum
                best_level = level

        return best_level

File: contests/lc/1161.MaximumLevelSumofaBinaryTree/solution.go
package main

// TreeNode is the standard LeetCode binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func maxLevelSum(root *TreeNode) int {
	if root == nil {
		return 0
	}

	queue := []*TreeNode{root}
	level := 0
	bestLevel := 1
	bestSum := root.Val

	for len(queue) > 0 {
		level++
		levelSum := 0
		next := make([]*TreeNode, 0, len(queue)*2)
		for _, node := range queue {
			levelSum += node.Val
			if node.Left != nil {
				next = append(next, node.Left)
			}
			if node.Right != nil {
				next = append(next, node.Right)
			}
		}

		if levelSum > bestSum || level == 1 {
			bestSum = levelSum
			bestLevel = level
		}
		queue = next
	}

	return bestLevel
}
